cxx/hirm.cc

```diff
@@ -3,16 +3,38 @@
 
 #include "hirm.hh"
 
-HIRM::HIRM(const T_schema& schema, std::mt19937* prng) {
-  // Add the clean relations before the noisy relations.
-  for (const auto& [name, relation] : schema) {
-    if (std::holds_alternative<T_clean_relation>(relation)) {
-      this->add_relation(prng, name, relation);
+HIRM::HIRM(const T_schema& _schema, std::mt19937* prng) {
+  printf("in hirm\n");
+  while (true) {
+    bool added_a_relation = false;
+    bool have_pending_relations = false;
+    for (const auto& [name, relation] : _schema) {
+      if (schema.contains(name)) {
+        continue;
+      }
+
+      if (std::holds_alternative<T_clean_relation>(relation)) {
+        this->add_relation(prng, name, relation);
+        added_a_relation = true;
+        continue;
+      }
+
+      const std::string base_rel = std::get<T_noisy_relation>(relation).base_relation;
+      if (schema.contains(base_rel)) {
+        this->add_relation(prng, name, relation);
+        added_a_relation = true;
+      } else {
+        have_pending_relations = true;
+      }
     }
-  }
-  for (const auto& [name, relation] : schema) {
-    if (std::holds_alternative<T_noisy_relation>(relation)) {
-      this->add_relation(prng, name, relation);
+
+    if (!have_pending_relations) {
+      return;
+    }
+
+    if (!added_a_relation) {
+      printf("Detected loop in schema\n");
```

Comment:
I think it'd be helpful to print the relation name.

Reply:
This is outside of the loop over relations, so there is no single relation name to print.